pkg/admission/prep_downscale.go

```diff
@@ -0,0 +1,211 @@
+		}
+	default:
+		return allowWarn(logger, fmt.Sprintf("unsupported type %T, allowing the change", o))
+	}
+
+	if lbls[PrepDownscaleLabelKey] != PrepDownscaleLabelValue {
+		// Not labeled, nothing to do.
+		return &v1.AdmissionResponse{Allowed: true}
+	}
+
+	port, ok := lbls[PrepDownscalePortKey]
+	if !ok || len(port) == 0 {
+		reviewResponse := v1.AdmissionResponse{
+			Allowed: false,
+			Result: &metav1.Status{
+				Message: fmt.Sprintf("downscale of %s/%s in %s from %d to %d replicas is not allowed because the %v label is not set or empty.", ar.Request.Resource.Resource, ar.Request.Name, ar.Request.Namespace, *oldReplicas, *newReplicas, PrepDownscalePortKey),
+			},
+		}
+		level.Warn(logger).Log("msg", fmt.Sprintf("downscale not allowed because the %v label is not set or empty", PrepDownscalePortKey))
+		return &reviewResponse
+	}
+
+	path, ok := lbls[PrepDownscalePathKey]
+	if !ok || len(path) == 0 {
+		reviewResponse := v1.AdmissionResponse{
+			Allowed: false,
+			Result: &metav1.Status{
+				Message: fmt.Sprintf("downscale of %s/%s in %s from %d to %d replicas is not allowed because the %v label is not set or empty.", ar.Request.Resource.Resource, ar.Request.Name, ar.Request.Namespace, *oldReplicas, *newReplicas, PrepDownscalePathKey),
+			},
+		}
+		level.Warn(logger).Log("msg", fmt.Sprintf("downscale not allowed because the %v label is not set or empty", PrepDownscalePathKey))
+		return &reviewResponse
+	}
+
+	// Since it's a downscale, check if the resource has the label that indicates it needs to be prepared to be downscaled.
+	// Create a slice of endpoint addresses for pods to send HTTP post requests to and to fail if any don't return 200
+	if !*ar.Request.DryRun {
+		diff := (*oldReplicas - *newReplicas)
+		type endpoint struct {
+			url   string
+			index int
+		}
+		eps := make([]endpoint, diff)
+
+		// The DNS entry for a pod of a stateful set is
+		// ingester-zone-a-0.$(servicename).$(namespace).svc.cluster.local
+		// The service in this case is ingester-zone-a as well.
+		// https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#stable-network-id
+		for i := 0; i < int(diff); i++ {
+			index := int(*oldReplicas) - i - 1 // nr in statefulset
+			eps[i].url = fmt.Sprintf("%v-%v.%v.%v.svc.cluster.local:%s/ingester/%s",
+				ar.Request.Name, // pod name
+				index,
+				ar.Request.Name, // svc name
+				ar.Request.Namespace,
+				port,
+				path,
+			)
+			eps[i].index = index
+		}
+
+		g, _ := errgroup.WithContext(ctx)
+		for _, ep := range eps {
+			ep := ep // https://golang.org/doc/faq#closures_and_goroutines
+			g.Go(func() error {
+				logger = log.With(logger, "url", ep.url, "index", ep.index)
+				err = addPreparedForDownscaleAnnotationToPod(ctx, api, ar.Request.Namespace, ar.Request.Name, ep.index)
+				if err != nil {
+					level.Debug(logger).Log("msg", "adding annotation to pod failed", "err", err)
+				}
+				level.Debug(logger).Log("msg", "annotation added to pod")
```

Comment:
What is the purpose of annotating the pod? We can't even trust this annotation, as it will be removed if the pod is rescheduled.

Reply:
These are there for restarts of the pod. If this happens before the pod could be shutdown it will start up again without `prepare_shutdown` being called on it. https://github.com/grafana/loki/pull/9153 will use it's presence to call `prepare_shutdown`.